File: samples/java/TicTacToe/Client/src/main/java/systems/zlink/samples/tictactoe/client/TicTacToeClientScenario.java
package systems.zlink.samples.tictactoe.client;

import systems.zlink.httpclient.ZLinkHttpClient;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticateReq;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticateRes;
import systems.zlink.samples.tictactoe.shared.contracts.CreateGameHttpReq;
import systems.zlink.samples.tictactoe.shared.contracts.CreateGameHttpRes;
import systems.zlink.samples.tictactoe.shared.contracts.GameStateNotify;
import systems.zlink.samples.tictactoe.shared.contracts.JoinGameMsg;
import systems.zlink.samples.tictactoe.shared.contracts.JoinGameNotify;
import systems.zlink.samples.tictactoe.shared.contracts.LeaveGameMsg;
import systems.zlink.samples.tictactoe.shared.contracts.ObserveMilestoneReq;
import systems.zlink.samples.tictactoe.shared.contracts.ObserveMilestoneRes;
import systems.zlink.samples.tictactoe.shared.contracts.PlaceMarkReq;
import systems.zlink.samples.tictactoe.shared.contracts.PlaceMarkRes;
import systems.zlink.samples.tictactoe.shared.contracts.PlayerJoinedNotify;
import systems.zlink.samples.tictactoe.shared.contracts.WinMilestoneNotify;
import systems.zlink.stream.connector.ZLinkStreamConnector;
import systems.zlink.stream.connector.ZLinkStreamConnectorFactory;
import systems.zlink.stream.connector.ZLinkStreamConnectorOptions;
import systems.zlink.stream.connector.ZLinkStreamDispatchMode;

import java.net.URI;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardWatchEventKinds;
import java.nio.file.WatchKey;
import java.nio.file.WatchService;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;

public final class TicTacToeClientScenario {
    public void run(TicTacToeClientOptions options) throws Exception {
        CreateGameHttpRes game;
        try (ZLinkHttpClient api = ZLinkHttpClient.create(options.apiUrl()).build()) {
            game =
                    api.post("/games")
                            .body(new CreateGameHttpReq(options.gameName()))
                            .submit(CreateGameHttpRes.class)
                            .toCompletableFuture()
                            .join()
                            .body();
        }
        ensure(game.playEndpoints().size() >= 2);
        ZLinkStreamConnector host = playerConnector(game.playEndpoints().get(0), "host");
        ZLinkStreamConnector guest = playerConnector(game.playEndpoints().get(1), "guest");
        ZLinkStreamConnector observer = playerConnector(game.playEndpoints().get(1), "observer");
        ZLinkStreamConnector reconnectedHost = null;
        boolean hostClosed = false;

        try {
            host.connect().submit().toCompletableFuture().join();
            guest.connect().submit().toCompletableFuture().join();
            observer.connect().submit().toCompletableFuture().join();
            System.out.println("observer-connected endpoint=" + game.playEndpoints().get(1));

            ensure(game.roomId() != null && !game.roomId().isBlank());
            ensure(options.gameName().equals(game.gameName()));
            ensure(game.requiredLevel() == 3);

            AuthenticateRes hostAuth =
                    host.request(new AuthenticateReq(options.xActorId()))
                            .submit(AuthenticateRes.class)
                            .toCompletableFuture()
                            .join();
            ensure(options.xActorId().equals(hostAuth.player().actorId()));
            ensure(hostAuth.player().wins() == 99);

            AuthenticateRes guestAuth =
                    guest.request(new AuthenticateReq(options.oActorId()))
                            .submit(AuthenticateRes.class)
                            .toCompletableFuture()
                            .join();
            ensure(options.oActorId().equals(guestAuth.player().actorId()));
            ensure(!guestAuth.player().actorId().equals(hostAuth.player().actorId()));

            AuthenticateRes observerAuth =
                    observer.request(new AuthenticateReq(options.observerActorId()))
                            .submit(AuthenticateRes.class)
                            .toCompletableFuture()
                            .join();
            ensure(options.observerActorId().equals(observerAuth.player().actorId()));
            ObserveMilestoneRes subscribed =
                    observer.request(new ObserveMilestoneReq())
                            .submit(ObserveMilestoneRes.class)
                            .toCompletableFuture()
                            .join();
            ensure(subscribed.subscribed());
            System.out.println(
                    "observer-subscription=verified subscribed=" + subscribed.subscribed());

            AtomicInteger hostOwnJoinNotifications = new AtomicInteger();
            host.on(
                    PlayerJoinedNotify.class,
                    message -> {
                        if (options.xActorId().equals(message.payload().actorId())) {
                            hostOwnJoinNotifications.incrementAndGet();
                        }
                        return CompletableFuture.completedFuture(null);
                    });
            AtomicInteger guestOwnJoinNotifications = new AtomicInteger();
            guest.on(
                    PlayerJoinedNotify.class,
                    message -> {
                        if (options.oActorId().equals(message.payload().actorId())) {
                            guestOwnJoinNotifications.incrementAndGet();
                        }
                        return CompletableFuture.completedFuture(null);
                    });

            var hostJoinCompletion =
                    host.waitFor(JoinGameNotify.class).submit(JoinGameNotify.class);
            host.send(new JoinGameMsg(game.roomId())).submit().toCompletableFuture().join();
            JoinGameNotify hostJoin = hostJoinCompletion.toCompletableFuture().join().payload();
            ensure(hostJoin.state().roomId().equals(game.roomId()));
            ensure("WaitingForPlayers".equals(hostJoin.state().status()));
            ensure(options.xActorId().equals(hostJoin.state().xActorId()));

            var hostSawGuestJoin =
                    host.waitFor(PlayerJoinedNotify.class)
                            .where(
                                    PlayerJoinedNotify.class,
                                    message ->
                                            options.oActorId().equals(message.payload().actorId()))
                            .submit(PlayerJoinedNotify.class);
            var hostSawGameStart =
                    host.waitFor(GameStateNotify.class)
                            .where(
                                    GameStateNotify.class,
                                    message ->
                                            "InProgress".equals(message.payload().state().status())
                                                    && options.oActorId()
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .oActorId()))
                            .submit(GameStateNotify.class);

            var guestJoinCompletion =
                    guest.waitFor(JoinGameNotify.class).submit(JoinGameNotify.class);
            guest.send(new JoinGameMsg(game.roomId())).submit().toCompletableFuture().join();
            JoinGameNotify guestJoin = guestJoinCompletion.toCompletableFuture().join().payload();
            ensure(guestJoin.state().roomId().equals(game.roomId()));
            ensure("InProgress".equals(guestJoin.state().status()));
            ensure(options.oActorId().equals(guestJoin.state().oActorId()));

            PlayerJoinedNotify guestJoinNotify =
                    hostSawGuestJoin.toCompletableFuture().join().payload();
            ensure("O".equals(guestJoinNotify.mark()));
            ensure("InProgress".equals(guestJoinNotify.state().status()));

            GameStateNotify gameStart = hostSawGameStart.toCompletableFuture().join().payload();
            ensure("X".equals(gameStart.state().nextTurn()));

            var guestSawHostMove1 =
                    guest.waitFor(GameStateNotify.class)
                            .where(
                                    GameStateNotify.class,
                                    message ->
                                            options.xActorId()
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .lastMoveActorId())
                                                    && Integer.valueOf(0)
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .lastMoveCell()))
                            .submit(GameStateNotify.class);
            PlaceMarkRes hostMove1 =
                    host.request(new PlaceMarkReq(0))
                            .submit(PlaceMarkRes.class)
                            .toCompletableFuture()
                            .join();
            ensure("X........".equals(hostMove1.state().board()));
            ensure("O".equals(hostMove1.state().nextTurn()));
            ensure(options.xActorId().equals(hostMove1.state().lastMoveActorId()));
            ensure(Integer.valueOf(0).equals(hostMove1.state().lastMoveCell()));

            GameStateNotify hostMove1Notify =
                    guestSawHostMove1.toCompletableFuture().join().payload();
            ensure(hostMove1Notify.state().board().equals(hostMove1.state().board()));
            ensure("O".equals(hostMove1Notify.state().nextTurn()));
            ensure(options.xActorId().equals(hostMove1Notify.state().lastMoveActorId()));
            ensure(Integer.valueOf(0).equals(hostMove1Notify.state().lastMoveCell()));

            var hostSawGuestMove1 =
                    host.waitFor(GameStateNotify.class)
                            .where(
                                    GameStateNotify.class,
                                    message ->
                                            options.oActorId()
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .lastMoveActorId())
                                                    && Integer.valueOf(3)
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .lastMoveCell()))
                            .submit(GameStateNotify.class);
            PlaceMarkRes guestMove1 =
                    guest.request(new PlaceMarkReq(3))
                            .submit(PlaceMarkRes.class)
                            .toCompletableFuture()
                            .join();
            ensure("X..O.....".equals(guestMove1.state().board()));
            ensure("X".equals(guestMove1.state().nextTurn()));
            ensure(options.oActorId().equals(guestMove1.state().lastMoveActorId()));
            ensure(Integer.valueOf(3).equals(guestMove1.state().lastMoveCell()));

            GameStateNotify guestMove1Notify =
                    hostSawGuestMove1.toCompletableFuture().join().payload();
            ensure(guestMove1Notify.state().board().equals(guestMove1.state().board()));
            ensure("X".equals(guestMove1Notify.state().nextTurn()));
            ensure(options.oActorId().equals(guestMove1Notify.state().lastMoveActorId()));
            ensure(Integer.valueOf(3).equals(guestMove1Notify.state().lastMoveCell()));

            var guestSawHostMove2 =
                    guest.waitFor(GameStateNotify.class)
                            .where(
                                    GameStateNotify.class,
                                    message ->
                                            options.xActorId()
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .lastMoveActorId())
                                                    && Integer.valueOf(1)
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .lastMoveCell()))
                            .submit(GameStateNotify.class);
            PlaceMarkRes hostMove2 =
                    host.request(new PlaceMarkReq(1))
                            .submit(PlaceMarkRes.class)
                            .toCompletableFuture()
                            .join();
            ensure("XX.O.....".equals(hostMove2.state().board()));
            ensure("O".equals(hostMove2.state().nextTurn()));
            ensure(options.xActorId().equals(hostMove2.state().lastMoveActorId()));
            ensure(Integer.valueOf(1).equals(hostMove2.state().lastMoveCell()));

            GameStateNotify hostMove2Notify =
                    guestSawHostMove2.toCompletableFuture().join().payload();
            ensure(hostMove2Notify.state().board().equals(hostMove2.state().board()));
            ensure("O".equals(hostMove2Notify.state().nextTurn()));
            ensure(options.xActorId().equals(hostMove2Notify.state().lastMoveActorId()));
            ensure(Integer.valueOf(1).equals(hostMove2Notify.state().lastMoveCell()));

            var hostSawGuestMove2 =
                    host.waitFor(GameStateNotify.class)
                            .where(
                                    GameStateNotify.class,
                                    message ->
                                            options.oActorId()
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .lastMoveActorId())
                                                    && Integer.valueOf(4)
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .lastMoveCell()))
                            .submit(GameStateNotify.class);
            PlaceMarkRes guestMove2 =
                    guest.request(new PlaceMarkReq(4))
                            .submit(PlaceMarkRes.class)
                            .toCompletableFuture()
                            .join();
            ensure("XX.OO....".equals(guestMove2.state().board()));
            ensure("X".equals(guestMove2.state().nextTurn()));
            ensure(options.oActorId().equals(guestMove2.state().lastMoveActorId()));
            ensure(Integer.valueOf(4).equals(guestMove2.state().lastMoveCell()));

            GameStateNotify guestMove2Notify =
                    hostSawGuestMove2.toCompletableFuture().join().payload();
            ensure(guestMove2Notify.state().board().equals(guestMove2.state().board()));
            ensure("X".equals(guestMove2Notify.state().nextTurn()));
            ensure(options.oActorId().equals(guestMove2Notify.state().lastMoveActorId()));
            ensure(Integer.valueOf(4).equals(guestMove2Notify.state().lastMoveCell()));

            var guestSawHostWin =
                    guest.waitFor(GameStateNotify.class)
                            .where(
                                    GameStateNotify.class,
                                    message ->
                                            "Won".equals(message.payload().state().status())
                                                    && options.xActorId()
                                                            .equals(
                                                                    message.payload()
                                                                            .state()
                                                                            .winner()))
                            .submit(GameStateNotify.class);
            var observerSawMilestone =
                    observer.waitFor(WinMilestoneNotify.class)
                            .where(
                                    WinMilestoneNotify.class,
                                    message ->
                                            options.xActorId().equals(message.payload().actorId())
                                                    && message.payload().wins() == 100)
                            .submit(WinMilestoneNotify.class);
            PlaceMarkRes hostWin =
                    host.request(new PlaceMarkReq(2))
                            .submit(PlaceMarkRes.class)
                            .toCompletableFuture()
                            .join();
            ensure("XXXOO....".equals(hostWin.state().board()));
            ensure("Won".equals(hostWin.state().status()));
            ensure(options.xActorId().equals(hostWin.state().winner()));

            GameStateNotify hostWinNotify = guestSawHostWin.toCompletableFuture().join().payload();
            ensure(hostWinNotify.state().board().equals(hostWin.state().board()));
            ensure("Won".equals(hostWinNotify.state().status()));
            ensure(options.xActorId().equals(hostWinNotify.state().winner()));

            WinMilestoneNotify milestone =
                    observerSawMilestone.toCompletableFuture().join().payload();
            ensure(milestone.wins() == 100);
            ensure(hostOwnJoinNotifications.get() == 0);
            ensure(guestOwnJoinNotifications.get() == 0);
            System.out.println(
                    "observer-win-milestone=verified actor="
                            + milestone.actorId()
                            + " wins="
                            + milestone.wins());

            // Closing this connector removes the session binding. A distinct
            // connector must authenticate and bind the existing Actor again.
            host.close().submit().toCompletableFuture().join();
            hostClosed = true;
            reconnectedHost = playerConnector(game.playEndpoints().get(0), "reconnected-host");
            reconnectedHost.connect().submit().toCompletableFuture().join();
            AuthenticateRes reconnectedAuth =
                    reconnectedHost
                            .request(new AuthenticateReq(options.xActorId()))
                            .submit(AuthenticateRes.class)
                            .toCompletableFuture()
                            .join();
            ensure(reconnectedAuth.player().equals(hostAuth.player()));

            // JoinGameMsg stays one-way on reconnect. Register the public
            // wait before send so a fast current-session push cannot be missed.
            var reconnectedJoinCompletion =
                    reconnectedHost
                            .waitFor(JoinGameNotify.class)
                            .where(
                                    JoinGameNotify.class,
                                    message ->
                                            game.roomId()
                                                    .equals(message.payload().state().roomId()))
                            .submit(JoinGameNotify.class);
            reconnectedHost
                    .send(new JoinGameMsg(game.roomId()))
                    .submit()
                    .toCompletableFuture()
                    .join();
            JoinGameNotify reconnectedJoin =
                    reconnectedJoinCompletion.toCompletableFuture().join().payload();
            ensure(reconnectedJoin.state().equals(hostWin.state()));
            System.out.println(
                    "reconnected-game-state=verified actor="
                            + reconnectedAuth.player().actorId()
                            + " room="
                            + game.roomId());

            // LeaveGameMsg is one-way. Keep the connectors open until the runner
            // confirms the separate leave and Entry Spot destroy lifecycle evidence.
            reconnectedHost
                    .send(new LeaveGameMsg(game.roomId()))
                    .submit()
                    .toCompletableFuture()
                    .join();
            guest.send(new LeaveGameMsg(game.roomId())).submit().toCompletableFuture().join();
            waitForLifecycleCompletion(options.lifecycleCompletionFile());
        } finally {
            if (!hostClosed) {
                host.close().submit().toCompletableFuture().join();
            }
            if (reconnectedHost != null) {
                reconnectedHost.close().submit().toCompletableFuture().join();
            }
            guest.close().submit().toCompletableFuture().join();
            observer.close().submit().toCompletableFuture().join();
        }
    }

    private static void waitForLifecycleCompletion(String completionFile) throws Exception {
        if (completionFile.isBlank()) {
            return;
        }

        Path releaseFile = Path.of(completionFile).toAbsolutePath().normalize();
        long deadlineNanos = System.nanoTime() + Duration.ofSeconds(60).toNanos();
        try (WatchService watcher = FileSystems.getDefault().newWatchService()) {
            releaseFile.getParent().register(watcher, StandardWatchEventKinds.ENTRY_CREATE);
            if (Files.exists(releaseFile)) {
                return;
            }

            long remainingNanos;
            while ((remainingNanos = deadlineNanos - System.nanoTime()) > 0) {
                WatchKey key =
                        watcher.poll(remainingNanos, java.util.concurrent.TimeUnit.NANOSECONDS);
                if (key == null) {
                    break;
                }
                boolean completionCreated =
                        key.pollEvents().stream()
                                .anyMatch(
                                        event -> releaseFile.getFileName().equals(event.context()));
                boolean valid = key.reset();
                if ((completionCreated || !valid) && Files.exists(releaseFile)) {
                    return;
                }
            }
            if (Files.exists(releaseFile)) {
                return;
            }
        }
        throw new IllegalStateException("Timed out waiting for runner lifecycle completion.");
    }

    private static ZLinkStreamConnector playerConnector(String endpoint, String role) {
        ZLinkStreamConnector connector =
                ZLinkStreamConnectorFactory.create(
                        new ZLinkStreamConnectorOptions(
                                URI.create(endpoint),
                                ZLinkStreamDispatchMode.IMMEDIATE,
                                TicTacToeSampleDefaults.RequestTimeout,
                                2,
                                Duration.ofSeconds(5),
                                64 * 1024,
                                false,
                                Duration.ofSeconds(1),
                                TicTacToeSampleDefaults.RequestTimeout.plusSeconds(5),
                                true,
                                Duration.ofMillis(250),
                                Duration.ofSeconds(5),
                                2.0,
                                null));
        return connector;
    }

    private static void ensure(boolean condition) {
        if (!condition) {
            throw new IllegalStateException("Ensure failed");
        }
    }
}

File: samples/kotlin/TicTacToe/Client/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/client/TicTacToeClientScenario.kt
package systems.zlink.samples.kotlin.tictactoe.client

import java.net.URI
import java.nio.file.Files
import java.nio.file.Path
import java.time.Duration
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import systems.zlink.framework.kotlin.ZLinkKotlinStreamConnector
import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.awaitReply
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.httpclient.kotlin.fetch
import systems.zlink.httpclient.kotlin.zlinkHttpClient
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticateReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticateRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.CreateGameHttpReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.CreateGameHttpRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.GameStateNotify
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.JoinGameMsg
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.JoinGameNotify
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.LeaveGameMsg
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.ObserveMilestoneReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.ObserveMilestoneRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlaceMarkReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlaceMarkRes
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.PlayerJoinedNotify
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.WinMilestoneNotify
import systems.zlink.stream.connector.ZLinkStreamConnectorFactory
import systems.zlink.stream.connector.ZLinkStreamConnectorOptions
import systems.zlink.stream.connector.ZLinkStreamDispatchMode
import systems.zlink.stream.connector.ZLinkStreamJson

class TicTacToeClientScenario {
    suspend fun run(options: TicTacToeClientOptions) = coroutineScope {
        val game =
            zlinkHttpClient(options.apiUrl).use { api ->
                api.post("/games")
                    .body(CreateGameHttpReq(options.gameName))
                    .fetch<CreateGameHttpRes>()
            }
        ensure(game.playEndpoints.size >= 2)
        val hostStream = playerConnector(game.playEndpoints[0])
        val guestStream = playerConnector(game.playEndpoints[1])
        val observerStream = playerConnector(game.playEndpoints[1])
        var reconnectedHostStream: ZLinkKotlinStreamConnector? = null
        var hostClosed = false

        try {
            hostStream.connect().await()
            guestStream.connect().await()
            observerStream.connect().await()

            ensure(game.roomId.isNotBlank())
            ensure(game.gameName == options.gameName)
            ensure(game.requiredLevel == 3)
            ensure(game.playEndpoints.size >= 2)
            ensure(game.playEndpoints.distinct().size == game.playEndpoints.size)
            ensure(game.playNodes.size == game.playEndpoints.size)
            ensure(game.playNodes.map { it.streamEndpoint }.toSet() == game.playEndpoints.toSet())

            val xAuthentication =
                hostStream.request(AuthenticateReq(options.xActorId)).awaitReply<AuthenticateRes>()
            ensure(xAuthentication.player.actorId == options.xActorId)
            ensure(xAuthentication.player.displayName.isNotBlank())
            ensure(xAuthentication.player.level >= game.requiredLevel)
            ensure(xAuthentication.player.wins == 99)

            val oAuthentication =
                guestStream.request(AuthenticateReq(options.oActorId)).awaitReply<AuthenticateRes>()
            ensure(oAuthentication.player.actorId == options.oActorId)
            ensure(oAuthentication.player.actorId != xAuthentication.player.actorId)
            ensure(oAuthentication.player.displayName.isNotBlank())
            ensure(oAuthentication.player.level >= game.requiredLevel)

            val observerAuthentication =
                observerStream
                    .request(AuthenticateReq(options.observerActorId))
                    .awaitReply<AuthenticateRes>()
            ensure(observerAuthentication.player.actorId == options.observerActorId)
            println("observer-connected endpoint=${game.playEndpoints[1]}")
            val subscription =
                observerStream.request(ObserveMilestoneReq()).awaitReply<ObserveMilestoneRes>()
            ensure(subscription.subscribed)
            println("observer-subscription=verified subscribed=${subscription.subscribed}")

            val hostNoSelfJoin =
                async(start = CoroutineStart.UNDISPATCHED) {
                    hostStream
                        .expectNone<PlayerJoinedNotify>("PlayerJoinedNotify")
                        .within(Duration.ofMillis(400))
                        .await()
                }
            val xJoinWait =
                hostStream
                    .waitFor<JoinGameNotify>()
                    .where { message -> message.payload().state.roomId == game.roomId }
                    .let { wait -> async(start = CoroutineStart.UNDISPATCHED) { wait.await() } }
            hostStream.send(JoinGameMsg(game.roomId)).await()
            val xJoin = xJoinWait.await().payload()
            ensure(xJoin.state.roomId == game.roomId)
            ensure(xJoin.state.status == "WaitingForPlayers")
            ensure(xJoin.state.xActorId == options.xActorId)
            hostNoSelfJoin.await()

            val hostSawGuestJoin =
                hostStream
                    .waitFor<PlayerJoinedNotify>()
                    .where { message -> message.payload().actorId == options.oActorId }
                    .let { wait -> async { wait.await() } }
            val hostSawGameStart =
                hostStream
                    .waitFor<GameStateNotify>()
                    .where { message -> message.payload().state.status == "InProgress" }
                    .let { wait -> async { wait.await() } }

            val guestNoSelfJoin =
                async(start = CoroutineStart.UNDISPATCHED) {
                    guestStream
                        .expectNone<PlayerJoinedNotify>("PlayerJoinedNotify")
                        .within(Duration.ofMillis(400))
                        .await()
                }
            val oJoinWait =
                guestStream
                    .waitFor<JoinGameNotify>()
                    .where { message -> message.payload().state.roomId == game.roomId }
                    .let { wait -> async(start = CoroutineStart.UNDISPATCHED) { wait.await() } }
            guestStream.send(JoinGameMsg(game.roomId)).await()
            val oJoin = oJoinWait.await().payload()
            ensure(oJoin.state.roomId == game.roomId)
            ensure(oJoin.state.status == "InProgress")
            ensure(oJoin.state.oActorId == options.oActorId)

            val guestJoinNotify = hostSawGuestJoin.await().payload()
            ensure(guestJoinNotify.actorId == options.oActorId)
            ensure(guestJoinNotify.displayName == oAuthentication.player.displayName)
            ensure(guestJoinNotify.level == oAuthentication.player.level)
            ensure(guestJoinNotify.mark == "O")
            ensure(guestJoinNotify.roomId == game.roomId)
            ensure(guestJoinNotify.state.status == "InProgress")
            guestNoSelfJoin.await()

            val gameStarted = hostSawGameStart.await().payload()
            ensure(gameStarted.state.nextTurn == "X")

            val guestSawHostMove1 =
                guestStream
                    .waitFor<GameStateNotify>()
                    .where { message -> message.payload().state.lastMoveCell == 0 }
                    .let { wait -> async { wait.await() } }
            val hostMove1 = hostStream.request(PlaceMarkReq(0)).awaitReply<PlaceMarkRes>()
            ensure(hostMove1.state.board == "X........")
            ensure(hostMove1.state.nextTurn == "O")
            ensure(hostMove1.state.lastMoveActorId == options.xActorId)
            ensure(hostMove1.state.lastMoveCell == 0)

            val hostMove1Notify = guestSawHostMove1.await().payload()
            ensure(hostMove1Notify.state.board == hostMove1.state.board)
            ensure(hostMove1Notify.state.nextTurn == "O")
            ensure(hostMove1Notify.state.lastMoveActorId == options.xActorId)
            ensure(hostMove1Notify.state.lastMoveCell == 0)

            val hostSawGuestMove1 =
                hostStream
                    .waitFor<GameStateNotify>()
                    .where { message -> message.payload().state.lastMoveCell == 3 }
                    .let { wait -> async { wait.await() } }

            val guestMove1 = guestStream.request(PlaceMarkReq(3)).awaitReply<PlaceMarkRes>()
            ensure(guestMove1.state.board == "X..O.....")
            ensure(guestMove1.state.nextTurn == "X")
            ensure(guestMove1.state.lastMoveActorId == options.oActorId)
            ensure(guestMove1.state.lastMoveCell == 3)

            val guestMove1Notify = hostSawGuestMove1.await().payload()
            ensure(guestMove1Notify.state.board == guestMove1.state.board)
            ensure(guestMove1Notify.state.nextTurn == "X")
            ensure(guestMove1Notify.state.lastMoveActorId == options.oActorId)
            ensure(guestMove1Notify.state.lastMoveCell == 3)

            val guestSawHostMove2 =
                guestStream
                    .waitFor<GameStateNotify>()
                    .where { message -> message.payload().state.lastMoveCell == 1 }
                    .let { wait -> async { wait.await() } }
            val hostMove2 = hostStream.request(PlaceMarkReq(1)).awaitReply<PlaceMarkRes>()
            ensure(hostMove2.state.board == "XX.O.....")
            ensure(hostMove2.state.nextTurn == "O")
            ensure(hostMove2.state.lastMoveActorId == options.xActorId)
            ensure(hostMove2.state.lastMoveCell == 1)

            val hostMove2Notify = guestSawHostMove2.await().payload()
            ensure(hostMove2Notify.state.board == hostMove2.state.board)
            ensure(hostMove2Notify.state.nextTurn == "O")
            ensure(hostMove2Notify.state.lastMoveActorId == options.xActorId)
            ensure(hostMove2Notify.state.lastMoveCell == 1)

            val hostSawGuestMove2 =
                hostStream
                    .waitFor<GameStateNotify>()
                    .where { message -> message.payload().state.lastMoveCell == 4 }
                    .let { wait -> async { wait.await() } }
            val guestMove2 = guestStream.request(PlaceMarkReq(4)).awaitReply<PlaceMarkRes>()
            ensure(guestMove2.state.board == "XX.OO....")
            ensure(guestMove2.state.nextTurn == "X")
            ensure(guestMove2.state.lastMoveActorId == options.oActorId)
            ensure(guestMove2.state.lastMoveCell == 4)

            val guestMove2Notify = hostSawGuestMove2.await().payload()
            ensure(guestMove2Notify.state.board == guestMove2.state.board)
            ensure(guestMove2Notify.state.nextTurn == "X")
            ensure(guestMove2Notify.state.lastMoveActorId == options.oActorId)
            ensure(guestMove2Notify.state.lastMoveCell == 4)

            val guestSawHostWin =
                guestStream
                    .waitFor<GameStateNotify>()
                    .where { message -> message.payload().state.status == "Won" }
                    .let { wait -> async { wait.await() } }
            val observerSawMilestone =
                observerStream
                    .waitFor<WinMilestoneNotify>()
                    .where { message ->
                        message.payload().actorId == options.xActorId &&
                            message.payload().wins == 100
                    }
                    .let { wait -> async { wait.await() } }
            val hostWin = hostStream.request(PlaceMarkReq(2)).awaitReply<PlaceMarkRes>()
            ensure(hostWin.state.board == "XXXOO....")
            ensure(hostWin.state.status == "Won")
            ensure(hostWin.state.winner == options.xActorId)

            val hostWinNotify = guestSawHostWin.await().payload()
            ensure(hostWinNotify.state.board == hostWin.state.board)
            ensure(hostWinNotify.state.status == "Won")
            ensure(hostWinNotify.state.winner == options.xActorId)

            val milestone = observerSawMilestone.await().payload()
            ensure(milestone.actorId == xAuthentication.player.actorId)
            ensure(milestone.displayName == xAuthentication.player.displayName)
            ensure(milestone.wins == 100)
            ensure(milestone.roomId == game.roomId)
            println(
                "observer-win-milestone=verified actor=${milestone.actorId} " +
                    "wins=${milestone.wins}"
            )

            // Closing this connector removes the session binding. A distinct
            // connector must authenticate and bind the existing Actor again.
            hostStream.close().await()
            hostClosed = true
            val freshHostStream = playerConnector(game.playEndpoints[0])
            reconnectedHostStream = freshHostStream
            freshHostStream.connect().await()
            val reconnectedAuthentication =
                freshHostStream
                    .request(AuthenticateReq(options.xActorId))
                    .awaitReply<AuthenticateRes>()
            ensure(reconnectedAuthentication.player == xAuthentication.player)

            // JoinGameMsg stays one-way on reconnect. Start the public wait
            // before send so a fast current-session push cannot be missed.
            val reconnectedJoinWait =
                freshHostStream
                    .waitFor<JoinGameNotify>()
                    .where { message -> message.payload().state.roomId == game.roomId }
                    .let { wait -> async(start = CoroutineStart.UNDISPATCHED) { wait.await() } }
            freshHostStream.send(JoinGameMsg(game.roomId)).await()
            val reconnectedJoin = reconnectedJoinWait.await().payload()
            ensure(reconnectedJoin.state == hostWin.state)
            println(
                "reconnected-game-state=verified actor=${reconnectedAuthentication.player.actorId} " +
                    "room=${game.roomId}"
            )

            // LeaveGameMsg remains one-way. Keep the connectors open until the runner
            // confirms the separate leave and Entry Spot destroy lifecycle evidence.
            freshHostStream.send(LeaveGameMsg(game.roomId)).await()
            guestStream.send(LeaveGameMsg(game.roomId)).await()
            waitForLifecycleCompletion(options.lifecycleCompletionFile)
        } finally {
            if (!hostClosed) {
                hostStream.close().await()
            }
            reconnectedHostStream?.let { it.close().await() }
            guestStream.close().await()
            observerStream.close().await()
        }
    }

    private fun playerConnector(endpoint: String): ZLinkKotlinStreamConnector =
        ZLinkStreamConnectorFactory.create(
                ZLinkStreamConnectorOptions(
                    URI.create(endpoint),
                    ZLinkStreamDispatchMode.IMMEDIATE,
                    TicTacToeSampleDefaults.RequestTimeout,
                    2,
                    Duration.ofSeconds(5),
                    64 * 1024,
                    false,
                    Duration.ofSeconds(1),
                    TicTacToeSampleDefaults.RequestTimeout.plusSeconds(5),
                    true,
                    Duration.ofMillis(250),
                    Duration.ofSeconds(5),
                    2.0,
                    ZLinkStreamJson.codec(),
                )
            )
            .kotlin()

    private suspend fun waitForLifecycleCompletion(completionFile: String) {
        if (completionFile.isBlank()) {
            return
        }

        val releaseFile = Path.of(completionFile).toAbsolutePath().normalize()
        val deadlineNanos = System.nanoTime() + Duration.ofSeconds(60).toNanos()
        while (System.nanoTime() < deadlineNanos) {
            if (Files.exists(releaseFile)) {
                return
            }
            delay(100)
        }
        error("Timed out waiting for runner lifecycle completion.")
    }
}

private fun ensure(condition: Boolean) {
    if (!condition) {
        throw IllegalStateException("Ensure failed")
    }
}
